### DISCUSSION TITLE
j4-dmenu-desktop + crazy-complete

### DISCUSSION THREAD

**Discussion by meator** [crazy-complete/crazy-complete#12] [Show and tell]
I'm happy to announce that crazy-complete is now completely integrated into j4-dmenu-desktop! It is currently present in the `develop` branch and it will be included in the upcoming `r3.2` release.

It can be found here: https://github.com/enkore/j4-dmenu-desktop/tree/develop/etc

I have fully integrated crazy-complete generation into j4-dmenu-desktop's build system. I plan to release generated completions alongside j4-dmenu-desktop's releases. crazy-complete isn't really packaged anywhere now, so I do not want to impose the requirement of installing crazy-complete on my users.

I also don't view including generated completions in the git repository as a good solution.

By the way, it might be worth including `j4-dmenu-desktop` in the "Definition file examples" section of the README. Real world examples of projects using crazy-complete may be useful for new users as a learning resource.

Apart from the issues I have already reported, I had no problems integrating crazy-complete into j4-dmenu-desktop. crazy-complete's syntax is capable of expressing j4-dmenu-desktop's flags comprehensively. j4-dmenu-desktop doesn't do overly complex argument handling and it doesn't do subcommands, but its flag handling is complicated enough for it to not be simple to represent in some competing shell completion generators.

**Comment by braph** [comment]
Hello Meator!

I'm really glad to see that `crazy-complete` is useful to you and has integrated so well into `j4-dmenu-desktop`!

But why not consider shipping the generated scripts within your git repository? This way, it would be much easier for users to install the completion files using your build system.

I'll definitely link to your project in my completions folder so others can see a practical example in action.

By the way, I want to say that working with you is a joy!

**Reply by meator** [reply]
Generated files should generally not appear in a git repository.

If I would add the generated files in, then it would be easy to have the source and the result files out of sync by accident (I could forget to update the generated completions after I modify the `.yaml` file). This cannot happen when completions are generated on the fly.

Users could also question whether the generated file was indeed generated by `crazy-complete`. I might have added some malicious code into generated completions that isn't present in the source file. It is possible to verify that the generated file was created by `crazy-complete`, but it's not that simple. If some other user generates the completion and compares their completion with the pregenerated one present in the repo, they may not match even though the pregenerated completion is legit. This may happen because the user was using a different version of `crazy-complete` than me when I generated them. Or they may have used the same version, but it still generated a different file, because their system configuration is different than mine (this is unlikely to happen with `crazy-complete`, but it is a problem for generator utilities which can be influenced by the execution environment they are in (env variables, files in `/etc/` and other directories...)).

Pregenerated completions also take up unnecessary space & git history. This isn't much of a problem for `crazy-complete`, because the generated files only occupy a couple of kilobytes.

**Comment by braph** [comment]
Hello Meator!

I'm announcing a new version of `crazy-complete`: **0.2.0**

What has changed?

The `group` parameter has been renamed to `groups` and now takes a list of groups instead of a single group.

I think the reasons why this change was necassary is quite obvious: Because sometimes options need to belong to multiple mutually-exclusive groups.

**Reply by meator** [reply]
I have updated j4-dmenu-desktop's completions: https://github.com/enkore/j4-dmenu-desktop/commit/830a1c733ea8fe92e431a8490aeb8cd7a8fbf39e

**Comment by braph** [comment]
Hi Meator,

I released version **0.3.0** of crazy-complete.

- The bash completion has changed. Options with arguments are now suffixed by "=" while completing them. This way, the user can see if an option is followed by an argument. (I noticed that the old behaviour was not consistent with other completion scripts used in `bash-completion`)
- The option parameter `multiple_option` has been renamed to `repeatable`
- Added backwards compatibility support for the old parameters. A deprecation warning will be printed if `group` or `multiple_option` is used. I will keep this strategy wherever possible.
- The shebang in the Debian package has been changed to `#/usr/bin/python3` (btw, which distro do you use? Is there any need for a new package?)

I'm also thinking of a new feature: An option parameter "hide", "hidden" or "suppress" (which one is best? your opinion?). This parameter would disable the option in the completion process, but arguments for those options could still be completed if the option is specified.

**Reply by meator** [reply]
Earlier today, I have [added a warning into j4-dmenu-desktop's build system](https://github.com/enkore/j4-dmenu-desktop/blob/2bcac73c38d41cc5843f4f2374d5c24a5db4f03f/etc/meson.build#L26-L30) that warns if a newer version of crazy-complete is used than the one I tested. I have implemented it because of stability status of crazy-complete described in #14. I didn't expect that such feature would find its use this soon.

I don't think it's that necessary to go all in with proper deprecation of old features now. I guess this is related to https://github.com/crazy-complete/crazy-complete/issues/14. This project is not stable yet (or that's how I interpreted your responses in #14), so deprecating things just adds complexity. It could be valuable to see how such deprecations would be handled after crazy-complete is declared stable. If you want to test that, leave the deprecation warnings be, but remove them when you stabilize crazy-complete. The 1.0.0 version should start with a "clean slate", all deprecated options should behave like any other invalid option for the first stable release (this is how I would handle it, but there isn't a one set approach for handling these things; the approach would of course be different if crazy-complete had a large user base before it is declared stable).

By the way, I can confirm that j4-dmenu-desktop's completions generate correctly with crazy-complete 0.3.0 even though I haven't updated them yet. The current deprecation system works as expected.

> * The shebang in the Debian package has been changed to `#/usr/bin/python3`

I haven't found any changes to shebangs with a brief look through the git log. What do you mean by this change?

I'm on Void Linux. I install crazy-complete through pip into a venv. I install a local git checkout of my desired version of crazy-complete.

> I'm also thinking of a new feature: An option parameter "hide", "hidden" or "suppress" (which one is best? your opinion?). This parameter would disable the option in the completion process, but arguments for those options could still be completed if the option is specified.

This is a pretty niche feature (but some of the features I have requested could also be considered niche, so I'm not complaining). I have noticed that some Python argument handlers support such "secret" flags that are handled by the completion system, but not shown in `--help`. I personally find this feature nonsensical, because `--help` will have to lie to the user about these flags. I'm partially biased here, because I had to debug a really annoying problem related to such hidden flag. I had to dig really deep to figure out that the flag I was searching for even exists!

What is the purpose of such hidden flags? Correct me if I'm wrong (I don't lie to my users about the flags my programs accept, so I haven't really used this feature before), but from my experience these flags are meant for internal use only (in service scripts, generated scripts not meant to be edited by users etc.). An example that comes to my mind is `meson --internal`. `meson --help` doesn't show it:

```
usage: meson [-h] {setup,configure,dist,install,introspect,init,test,wrap,subprojects,rewrite,compile,devenv,env2mfile,format,fmt,help} ...

options:
  -h, --help                                                                         show this help message and exit

Commands:
  If no command is specified it defaults to setup command.

  {setup,configure,dist,install,introspect,init,test,wrap,subprojects,rewrite,compile,devenv,env2mfile,format,fmt,help}
    setup                                                                            Configure the project
    configure                                                                        Change project options
    dist                                                                             Generate release archive
    install                                                                          Install the project
    introspect                                                                       Introspect project
    init                                                                             Create a new project
    test                                                                             Run tests
    wrap                                                                             Wrap tools
    subprojects                                                                      Manage subprojects
    rewrite                                                                          Modify the project definition
    compile                                                                          Build the project
    devenv                                                                           Run commands in developer environment
    env2mfile                                                                        Convert current environment to a cross or native file
    format (fmt)                                                                     Format meson source file
    help                                                                             Print help of a subcommand
```

It is used in `build.ninja` files generated by Meson. The contents of this file are an implementation detail of Meson, they may not be edited. Relying on any internal behavior present in `build.ninja` files is not supported by Meson. The `--internal` flag is used to enable some additional functionality in the `build.ninja` file. It is a private method of communication between Ninja and Meson.[^1]

I don't think it's wise to complete these flags in any way (even for arguments only). Completion scripts should be written for the "public API" of command flags. If a program author has decided that a flag should be hidden, than completion scripts should treat it as if it doesn't exist.

But if you have a different use case of these hidden flags in mind that I didn't think of, please correct me.

[^1]: I'm simplifying it a bit here. Meson can also output VS solution files etc.

**Comment by braph** [comment]
Hello @meator!

I recently published two new releases of `crazy-complete`.

While none of the bugfixes affects you, I recommend regenerating your completion scripts for two reasons:

- **Fish** has dropped support for using `?` as a wildcard in `case`.
  Older versions of `crazy-complete` used this wildcard in the `query` function, which makes them incompatible with newer Fish releases.
- The generated **Bash** script for `j4-dmenu-desktop` is now about a hundred lines shorter.

Changes relevant to your `j4-dmenu-desktop` **Bash script**:

- Improved code for parsing the commandline. It is much shorter now.
- The helper function `_j4-dmenu-desktop_compgen_w_replacement()` was removed
  and replaced by `compgen -W`, since all of your completions don't include any unsafe characters.
- Option string completion has been optimized: repetitive final option checks were consolidated into a single check.

  Old style:

  ```bash
    (( ! ${#OPT_display_binary} && ! ${#OPT_help} && ! ${#OPT_version} )) && opts+=(-b --display-binary)
    (( ! ${#OPT_display_binary_base} && ! ${#OPT_help} && ! ${#OPT_version} )) && opts+=(-f --display-binary-base)
    [...]
  ```

  New style:

  ```bash
    if (( ! ${#OPT_help[@]} && ! ${#OPT_version[@]} )); then
      (( ! ${#OPT_display_binary[@]} )) && opts+=(-b --display-binary)
      (( ! ${#OPT_display_binary_base[@]} )) && opts+=(-f --display-binary-base)
      [...]
  ```

Best regards,

Benjamin

P.S.: Im working on https://github.com/crazy-complete/crazy-complete/issues/13 . Maybe it will be finished this week :-)

**Reply by meator** [reply]
Thanks for the heads up! I didn't expect this level of support from you.

I've updated j4-dmenu-desktop's completions: https://github.com/enkore/j4-dmenu-desktop/commit/d678609a0467a5fad71410fe24c552622cfe9810
